File: CPP/CaesarCipher.cpp
#include <iostream>
using namespace std;

string caesarCipher(string text, int key) {
    string result = "";
    for (int i = 0; i < text.length(); i++) {
        if (text[i] >= 'A' && text[i] <= 'Z') {
            char c = (text[i] + key - 'A') % 26 + 'A';
            if (c < 'A') c += 26;
            result += c;
        } else if (text[i] >= 'a' && text[i] <= 'z') {
            char c = (text[i] + key - 'a') % 26 + 'a';
            if (c < 'a') c += 26;
            result += c;
        } else {
            result += text[i];
        }
    }
    return result;
}

string caesarDecipher(string text, int key) {
    return caesarCipher(text, -key);
}

int main() {
    cout << caesarCipher("ABCXYZabcxyz Cat!Dog^Mouse.", -3) << endl;
    cout << caesarCipher("ABCXYZabcxyz Cat!Dog^Mouse.", -29) << endl;
    cout << caesarCipher("ABCXYZabcxyz Cat!Dog^Mouse.", 3) << endl;
    cout << caesarCipher("ABCXYZabcxyz Cat!Dog^Mouse.", 29) << endl;
    
    cout << caesarDecipher("XYZUVWxyzuvw Zxq!Ald^Jlrpb.", -3) << endl;
    cout << caesarDecipher("XYZUVWxyzuvw Zxq!Ald^Jlrpb.", -29) << endl;
    cout << caesarDecipher("DEFABCdefabc Fdw!Grj^Prxvh.", 3) << endl;
    cout << caesarDecipher("DEFABCdefabc Fdw!Grj^Prxvh.", 29) << endl;
}

File: Python/CaesarCipher.py
def caesar_cipher(text: str, key: int) -> str:
    result = ""
    for char in text:
        if 'A' <= char <= 'Z':
            result += chr((ord(char) + key - ord('A')) % 26 + ord('A'))
        elif 'a' <= char <= 'z':
            result += chr((ord(char) + key - ord('a')) % 26 + ord('a'))
        else:
            result += char
    return result

print(caesar_cipher("ABCXYZabcxyz Cat!Dog^Mouse.", -3))
print(caesar_cipher("ABCXYZabcxyz Cat!Dog^Mouse.", -29))
print(caesar_cipher("ABCXYZabcxyz Cat!Dog^Mouse.", 3))
print(caesar_cipher("ABCXYZabcxyz Cat!Dog^Mouse.", 29))

def caesar_decipher(text: str, key: int) -> str:
    return caesar_cipher(text, -key)

print(caesar_decipher("XYZUVWxyzuvw Zxq!Ald^Jlrpb.", -3))
print(caesar_decipher("XYZUVWxyzuvw Zxq!Ald^Jlrpb.", -29))
print(caesar_decipher("DEFABCdefabc Fdw!Grj^Prxvh.", 3))
print(caesar_decipher("DEFABCdefabc Fdw!Grj^Prxvh.", 29))